Assert if padding mask type is unexpected (#86353)

zrphercule opened the pull request:
Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/86353

Fix the issue described in
https://github.com/pytorch/pytorch/issues/86120

Test Plan: buck test mode/opt caffe2/test:test_transformers -- test_train_with_long_type_pad

Differential Revision: D40129968

malfet commented on torch/nn/modules/transformer.py:
```diff
@@ -443,6 +448,11 @@ def forward(self, src: Tensor, src_mask: Optional[Tensor] = None,
         """
 
         # see Fig. 1 of https://arxiv.org/pdf/2002.04745v1.pdf
+        if src_key_padding_mask is not None:
+            if src_key_padding_mask.dtype != torch.bool:
+                if not torch.is_floating_point(src_key_padding_mask):
```
Please consider dismantling pyramid of doom by a bit
```suggestion
        if src_key_padding_mask is not None:
            _skpm_dtype = src_key_padding_mask.dtype
            if _skpm_dtype != torch.bool and not _skpm_dtype.is_floating_point:
```

malfet commented on torch/nn/modules/activation.py:
```diff
@@ -1062,6 +1061,11 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
             `batch_first` argument is ignored for unbatched inputs.
         """
         is_batched = query.dim() == 3
+        if key_padding_mask is not None:
+            if key_padding_mask.dtype != torch.bool:
+                if not torch.is_floating_point(key_padding_mask):
+                    raise AssertionError(
+                        "only bool and floating type of key_padding_mask is supported")
```
You are adding the same check thrice, which seems to warrant a convenience function
```
def _assert_if_tensor_is_not_float_or_bool(t: Optional[torch.Tensor], name: str) -> None:
    if t is None:
        return
   if t.dtype != torch.bool and not t.dtype.is_floating_point:
       raise AssertionError(f"only bool and floating types of {name} are supported")
```

zrphercule replied:
That is a good idea, but dont want to add new helper functions in this diff anymore in case it will break torch script :(

malfet reviewed (approved):
LGTM, but please consider the posted feedback

mikekgfb commented:
@pytorchbot merge

malfet commented:
Also, please change title, as this PR is not 1.13 cherry pick, it simply targets master.
Something like: `Assert if padding mask type is unexpected`?

malfet commented on torch/nn/modules/activation.py:
```diff
@@ -1062,6 +1061,11 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
         is_batched = query.dim() == 3
+        if key_padding_mask is not None:
+            _skpm_dtype = key_padding_mask.dtype
+            if _skpm_dtype != torch.bool and not torch.is_floating_point(key_padding_mask):
```
Why call a function rather than query the property directly?
```suggestion
            if _skpm_dtype != torch.bool and not _skpm_dtype.is_floating_point:
```

zrphercule replied:
@malfet I tagged you in phabricator. It is because some scripted tensors have really weird tensor dtype, like an "int" instead of torch.int. I pasted some links in pharicator, and msged you directly ;)

malfet commented on torch/nn/modules/activation.py:
```diff
@@ -1062,6 +1061,11 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
         """
         is_batched = query.dim() == 3
+        if key_padding_mask is not None:
+            _skpm_dtype = key_padding_mask.dtype
```
Variable name supposed to start with first letter of long variable name, i.e. for `key_padding_mask` is should be `_kpm_dtype`
```suggestion
            _kpm_dtype = key_padding_mask.dtype
```

zrphercule replied:
lol right

zrphercule committed:
Assert if padding mask type is unexpected (#87106)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/87106

Pull Request resolved: https://github.com/pytorch/pytorch/pull/86353

Fix the issue described in
https://github.com/pytorch/pytorch/issues/86120

Test Plan: buck test mode/opt caffe2/test:test_transformers -- test_train_with_long_type_pad

Reviewed By: malfet

Differential Revision: D40129968

fbshipit-source-id: d7530d8c4548b9d3a3ef8b9740c93bee000ed809

zrphercule commented:
@pytorchbot merge

ZainRizvi commented:
@pytorchmergebot merge
